### PR TITLE
feat(ffi): add FFI interface to get operators along with fields while performing expressions validation

### PR DESCRIPTION
**Pull request by Oyami-Srk**
This interface allows the caller to validate an expression string with schema and get its operators/fields without needing a Router instance.

KAG-5013

### CODE REVIEW

File: src/ffi.rs
Comment (dndx) [inline src/ffi.rs:703]: Since the validation result is used by a library, for these kind of "internal errors" that the user should not see, we should return a more structured error instead of the string, that way the caller knows what happened and can reallocate if desired.

I suggest changing the return of this FFI function to a `int` and define a couple of constants instead.

For errors that the caller can't do much and has to propagate to the user, a string error is okay.

**Review by dndx — CHANGES_REQUESTED** [review]

File: src/ffi.rs
Comment (Oyami-Srk) [reply src/ffi.rs:713]: Fixed, thanks.

File: src/ffi.rs
Comment (ProBrian) [inline src/ffi.rs:683]: Maybe we could wrap `parse`, `validate` and `add_to_counter` logic in a single function, so that it could be call for both `Router` and `expression_validate`. 
The reason I mention this is that there maybe data structure change for above functions(e.g. I may change `add_to_counter`'s parameter in my PR for indexed fields), wrapping the implementaion logic in a single function would bring the benefit that we don't need to change every caller's code in future.

Comment (ProBrian) [reply src/ffi.rs:683]: I just noticed that here we use `ast.add_to_counter`, and when Hai's PR merged, router will use `cir.add_to_counter`, there would be different. Maybe not easy to extract to general logic.

File: src/ffi.rs
Comment (ProBrian) [inline src/ffi.rs:826]: I think we'd better add test cases for `ATC_ROUTER_EXPRESSION_VALIDATE_FAILED` and `ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL`.

File: src/ffi.rs
Comment (Oyami-Srk) [reply src/ffi.rs:683]: Given that `validate` is provided by a trait and `AST` should always implement this (I don't think we want to perform the validation on other data structures.) and `parse` is very natural from `AST` from `String`, I think `parse` and `validate` should be considered very stable interfaces. Indeed calling `parse` and then `validate` is a widespread use case but I don't think we should add a new interface just calling them together.

However, I did realize that we should not depend on `add_to_counter` for this interface thanks for your comment. Because `counter` should be an internal underlying method to achieve something that should not be public as the interface. Hai's PR changed the `semantics::FieldCounter` to be implemented by `Cir` rather than `AST` because of that "internality". This is an interface change and not a data structure change.

I'd like to introduce two traits like `semantics::GetFields` and `semantics::GetOperators` and implement this for `AST` only. This may provide a more idiomatic way to get data from `AST`.

WDYT? @dndx @ADD-SP @ProBrian

File: src/ffi.rs
Comment (Oyami-Srk) [reply src/ffi.rs]: fixed.

File: src/ffi.rs
Comment (ProBrian) [reply src/ffi.rs:683]: > I'd like to introduce two traits like `semantics::GetFields` and `semantics::GetOperators` and implement this for `AST` only. This may provide a more idiomatic way to get data from `AST`.

Sounds good, that will make the FFI layer more clear and brief.

File: src/ffi.rs
Comment (ADD-SP) [inline src/ffi.rs:697]:
```suggestion
    if !fields_buf.is_null() {
        assert!(
          !fields_len.is_null() && ! fields_total.is_null(),
          "fields_len and fields_total must be non-null when fields_buf is non-null"
        )
```

Comment (Oyami-Srk) [reply src/ffi.rs:697]: This would not allow a null `fields_buf` with non-null `fields_len` or `fields_total` to get the required buffer length and total field count without a valid buffer. This behavior is indeed NOT encouraged (since a waste of validating and getting fields/operators is called) but I think it's necessary to provide that form of interface to make it more flexible and keep consistency to other interfaces like `get_fields`.

File: src/semantics.rs
Comment (ProBrian) [inline src/semantics.rs:217]: Are there more forthcoming PRs? I see this function not been used in ffi.rs.
Is it possible that we just define this like `fn get_fields(predicates: &Vec<&Predicate>) -> HashSet<String>`, so that we can reuse the `predicates` list we got from `get_predicates` function, and no need to invoke `self.get_predicates()` again?

File: src/semantics.rs
Comment (ProBrian) [inline src/semantics.rs:233]: same as above comment, this function seems not been used.

File: src/semantics.rs
Comment (Oyami-Srk) [reply src/semantics.rs:217]: I don't think `get_fields(predicates: &Vec<&Predicate>)` is a good interface since callers should not really care about what the `Predicate` is. Those two traits are indeed unused yet, but I'll make further changes to it.

Comment (Oyami-Srk) [reply src/semantics.rs]: Done with commit 2cf3480, please check to see if this could be resolved.

**Review by ProBrian — APPROVED** [review]

File: src/ffi.rs
Comment (javierguerragiraldez) [inline src/ffi.rs:652]: any reason to use `u8` for strings here, and `i8` elsewhere?

File: src/ffi.rs
Comment (Oyami-Srk) [reply src/ffi.rs:611]: Sure, thanks for the suggestion. All fields got `'\0'` for the terminator.

File: src/ffi.rs
Comment (Oyami-Srk) [reply src/ffi.rs:652]: According to the [document](https://doc.rust-lang.org/std/ffi/type.c_char.html), `ffi::c_char` is defined as `i8`. However, the document also mentioned that On modern architectures this type will always be either `i8` or `u8`.

Choosing `u8` here is just to keep consistency with Rust internal bytes represent like [`String::as_bytes`](https://doc.rust-lang.org/std/string/struct.String.html#method.as_bytes). 

P.S. I saw all the FFI interfaces in this project have no consistent way to represent `c_char`. E.g. `fields` is `u8` in `router_get_fields` and `i8` in `context_add_value`.

**Comment by Oyami-Srk** [comment]
Rebased with the latest main branch. Nothing else has been changed except `Cargo.toml`, `Cargo.lock` and `src/semantics.rs` since 2 weeks ago.

File: src/ffi.rs
Comment (dndx) [reply src/ffi.rs:652]: @Oyami-Srk In general, we tries to use `u8` for `char` whenever possible, except when dealing with char provided by LuaJIT.

This is because LuaJIT converts string to `const char *`, and this is a signed value. If we accept `u8` LuaJIT will complain when attempting to convert a string.

**Comment by Oyami-Srk** [comment]
@javierguerragiraldez Could you please approve this if there's no other concern in this Rust implementation?

File: src/ffi.rs
Comment (dndx) [inline src/ffi.rs:608]: I think we should just be optimistic here and always ask the user to pass in a buffer, 99% of the time reallocate will not be necessary, and we can avoid having to parse the expression twice which could be expensive (especially with Regexes).

**Review by dndx — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (Oyami-Srk) [reply Cargo.toml:36]: Done. Now feature flag `expr_validation` has been removed.

File: src/ast.rs
Comment (Oyami-Srk) [reply src/ast.rs:79]: Done.

File: src/ast.rs
Comment (Oyami-Srk) [reply src/ast.rs:199]: Done. Additionally, a private function to get predicates from `Expresison` has been moved under ffi module.

File: src/ffi.rs
Comment (Oyami-Srk) [reply src/ffi.rs]: Done.

File: src/ffi.rs
Comment (Oyami-Srk) [reply src/ffi.rs]: I've removed the ability to acquire the buffer length. Callers must pass a valid buffer if the calling is for getting the fields in use.
But I keep the `fields_buf` null-able because this interface is `expression_validate` and callers may not always want fields. This will allow us to do more refactoring on the Kong side with this interface. Like getting rid of validating expressions by inserting and then deleting them.

**Comment by Oyami-Srk** [comment]
Did a little improvement:

1. Break `ffi.rs` into module.
2. Move everything guarded by the feature flag `expr_validation` to `ffi:expression`.
3. **Remove feature flag `expr_validation`**.
4. Remove public interfaces `get_field` and `get_operator` for `Predicate`.
5. Remove the redundant error message when the buffer is too small.
6. Remove the ability to fetch field count and total length without a valid `field_buf`. (The `field_buf` is still null-able to allow doing validation only.)

@dndx Could you review it when you have a moment?
@javierguerragiraldez Could you please re-generate the C header and update [cgo compile flag](https://github.com/Kong/go-atc-router/pull/6/files#diff-146f7e1ec8fd2395b53645c09467eddc8d3d5f9cb100235e546a36bedfb5a770R4) as the feature flag `expr_validation` is removed, and only need feature `ffi`.

**Comment by javierguerragiraldez** [comment]
> @javierguerragiraldez Could you please re-generate the C header and update [cgo compile flag] as the feature flag `expr_validation` is removed, and only need feature `ffi`.

:+1:  done

File: src/ffi/expression.rs
Comment (dndx) [inline src/ffi/expression.rs:132]: This should be moved to "Safety" section, as "`atc` does not point to a valid C-style string" will be a safety violation and we can not guarantee a panic will happen.

File: src/ffi/expression.rs
Comment (dndx) [inline src/ffi/expression.rs:106]:
```suggestion
/// - `fields_buf_len`: a pointer to the length of `fields_buf`.
```

I was a bit confused by the name, if this refers to the size of `fields_buf` we should make `fields_buf` part of the name as well.

File: src/ffi/expression.rs
Comment (dndx) [inline src/ffi/expression.rs:107]:
```suggestion
/// - `fields_total`: a pointer for storing the total number of unique fields used in the expression.
```

File: src/ffi/expression.rs
Comment (dndx) [inline src/ffi/expression.rs:123]: Do we even care about `fields_total` when `ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL`? I'm not sure what can we do with this value if `fields_buf` is truncated.

File: src/ffi/expression.rs
Comment (dndx) [inline src/ffi/expression.rs:125]: Do we have to make so much fields optional?

Doing so makes the documentation hard to understand and number of combinations explode, and not sure if it even provides us much value. Should we just make everything required always instead?

File: src/ffi/expression.rs
Comment (dndx) [inline src/ffi/expression.rs:11]: Not a hard blocker for merge, but should we consider using an `iter()` method instead? I noticed all of the use cases for returned `predicates` are iterations, and an iterator might be a better interface.

**Review by dndx — CHANGES_REQUESTED** [review]

File: src/ffi/expression.rs
Comment (Oyami-Srk) [reply src/ffi/expression.rs]: Done.

File: src/ffi/expression.rs
Comment (Oyami-Srk) [reply src/ffi/expression.rs]: Done. Removed incorrect panic statement for `atc` due to the safety statements have alreay addressed `atc` pointer.

File: src/ffi/expression.rs
Comment (Oyami-Srk) [reply src/ffi/expression.rs]: The intend of optional fields `field_buf` and `operators` is to make this interface could validate expression without the overhead of getting elements from it. And hopefully be used by the [resty.router.router.validate](https://github.com/Kong/atc-router/blob/8c5d4f5dbdebead39c77e7cd7280e2a30eba4539/lib/resty/router/router.lua#L110).
I'm not updating `router.lua` because this PR is out of scope to do that.
But seemingly that `validate` method needs fields as well and it is not on the hot path. So it will be fine if we get `fields` and `operators` for every validating.

Comment (Oyami-Srk) [reply src/ffi/expression.rs]: I also removed the assertion on null since this is covered by Safety statement and will lead to UB if any of them are null.

File: src/ffi/expression.rs
Comment (Oyami-Srk) [reply src/ffi/expression.rs:123]: Currently, the implementation will not write any thing if buffer is too small. It will save all fields to be writen in a vector and always calcuate required length. This could be improved once the iterator is in use.

File: src/ffi/expression.rs
Comment (Oyami-Srk) [reply src/ffi/expression.rs:11]: Will do before this PR getting merged. Iteration interface will improve a little performance.

Comment (Oyami-Srk) [reply src/ffi/expression.rs]: Done. Now use a stack-based iterator. Although using Morris traversal could achieve this with more space efficiency (O(1) in space, O(n) in time), it is more intrusive on the tree structure itself and may require a raw pointer on the tree. I don't think it's worth it.

File: src/ffi/expression.rs
Comment (Oyami-Srk) [reply src/ffi/expression.rs]: Done.

**Review by ADD-SP — APPROVED** [review]
LGTM, just a small thing.

File: src/ffi/expression.rs
Comment (Oyami-Srk) [reply src/ffi/expression.rs:200]: Done.

File: src/ffi/context.rs
Comment (dndx) [inline src/ffi/context.rs:15]:
```suggestion
/// This function never returns an error, however, it can panic if memory allocation failed.
```

**Review by dndx — APPROVED** [review]

**Pull request merged**
